Lib/test/test_ssl.py

```diff
@@ -4767,23 +4782,34 @@ def call_after_accept(unused):
 
         ready_for_server_wrap_socket.set()
         server.join()
+
         wrap_error = server.wrap_error
-        self.assertEqual(b"", server.received_data)
-        self.assertIsInstance(wrap_error, OSError)  # All platforms.
-        self.non_linux_skip_if_other_okay_error(wrap_error)
-        self.assertIsInstance(wrap_error, ssl.SSLError)
-        self.assertIn("before TLS handshake with data", wrap_error.args[1])
-        self.assertIn("before TLS handshake with data", wrap_error.reason)
-        self.assertNotEqual(0, wrap_error.args[0])
-        self.assertIsNone(wrap_error.library, msg="attr must exist")
+        server.wrap_error = None
+        try:
+            self.assertEqual(b"", server.received_data)
+            self.assertIsInstance(wrap_error, OSError)  # All platforms.
+            self.non_linux_skip_if_other_okay_error(wrap_error)
+            self.assertIsInstance(wrap_error, ssl.SSLError)
+            self.assertIn("before TLS handshake with data", wrap_error.args[1])
+            self.assertIn("before TLS handshake with data", wrap_error.reason)
+            self.assertNotEqual(0, wrap_error.args[0])
+            self.assertIsNone(wrap_error.library, msg="attr must exist")
+        finally:
+            # gh-108342: Explicitly break the reference cycle
+            wrap_error = None
+            server = None
 
     def test_preauth_data_to_tls_client(self):
+        server_can_continue_with_wrap_socket = threading.Event()
         client_can_continue_with_wrap_socket = threading.Event()
 
         def call_after_accept(conn_to_client):
+            if not server_can_continue_with_wrap_socket.wait(support.SHORT_TIMEOUT):
+                print("ERROR: test client took too long")
+
             # This forces an immediate connection close via RST on .close().
             set_socket_so_linger_on_with_zero_timeout(conn_to_client)
-            conn_to_client.send(
+            conn_to_client.sendall(
```

Comment:
Why sendall?  This data is smaller than the smallest possible IP MTU so it is guaranteed to be a single packet after a single system call no matter what.  If it were large data and sendall were used to send multiple packets that'd run the risk of extending the timing window to not reproduce the bug. However unlikely in a localhost scenario with our blocking events in place.

The idea here is a single data packet followed by an RST.  Which is something user space socket APIs abstract away but these API calls happen to cause to happen on Linux.

Is this fixing an actual observable test issue?

Reply:
Well, while I was debugging this issue, I tried **many** different things to fix all race conditions :-) When I got a race condition on this function between the client and the server, I saw that the unsafe socket.send() function is used: it can do a partial send.

Sure, I can revert this change. I don't think that it's **needed**.